src/utilities/createRouter.ts

```diff
@@ -20,19 +21,42 @@ export function createRouter<const T extends Routes>(routes: T, options: RouterO
     const matched = routeMatch(resolved, url)
 
     if (!matched) {
-      await executeMiddleware(notFoundRouteResolved)
       return reject('NotFound')
     }
 
     try {
-      await executeMiddleware(matched)
-      updateRoute(matched)
-      clearRejection()
+      await executeMiddleware({
+        to: readonly(matched),
+        from: route,
+      })
     } catch (error) {
       if (error instanceof RouterRejectionError) {
         reject(error.type)
       }
+
+      if (error instanceof RouterPushError) {
+        const [source, options] = error.to
+        const url = resolve(source)
+
+        navigation.update(url, options)
+        return
+      }
+
+      if (error instanceof RouterReplaceError) {
+        const [source] = error.to
+        const url = resolve(source)
+
+        navigation.update(url, { replace: true })
+        return
+      }
```

Comment:
IMO moving this here makes a lot of sense. Leaves the router itself responsible for navigation.